availability-store/src/worker.rs

```diff
@@ -0,0 +1,989 @@
+				runtime_handle,
+				sender,
+				relay_parent,
+				candidate.erasure_root
+			) {
+				warn!(target: LOG_TARGET, "Failed to register chunk listener: {}", e);
+			}
+		}
+
+		let _ = self.availability_store.add_candidates_in_relay_block(
+			&relay_parent,
+			hashes
+		);
+
+		Ok(())
+	}
+
+	// Processes chunks messages that contain awaited items.
+	//
+	// When an awaited item is received, it is placed into the availability store
+	// and removed from the frontier. Listener de-registered.
+	fn on_chunks_received(
+		&mut self,
+		relay_parent: Hash,
+		candidate_hash: Hash,
+		chunks: Vec<ErasureChunk>
+	) -> Result<(), Error> {
+		let (_, n_validators) = self.availability_store
+			.get_validator_index_and_n_validators(&relay_parent)
+			.ok_or(Error::IdAndNValidatorsNotFound { relay_parent })?;
+
+		let receipt = self.availability_store.get_candidate(&candidate_hash)
+			.ok_or(Error::CandidateNotFound { candidate_hash })?;
+
+		for chunk in &chunks {
+			let topic = erasure_coding_topic(relay_parent, receipt.erasure_root, chunk.index);
+			// need to remove gossip listener and stop it.
+			if let Some(signal) = self.registered_gossip_streams.remove(&topic) {
+				signal.fire();
+			}
+		}
+
+		self.availability_store.add_erasure_chunks(
+			n_validators,
+			&relay_parent,
+			&candidate_hash,
+			chunks,
+		)?;
+
+		Ok(())
+	}
+
+	// Adds the erasure roots into the store.
+	fn on_erasure_roots_received(
+		&mut self,
+		relay_parent: Hash,
+		erasure_roots: Vec<Hash>
+	) -> Result<(), Error> {
+		self.availability_store.add_erasure_roots_in_relay_block(&relay_parent, erasure_roots)?;
+
+		Ok(())
+	}
+
+	// Processes the `ListenForChunks` message.
+	//
+	// When the worker receives a `ListenForChunk` message, it double-checks that
+	// we don't have that piece, and then it registers a listener.
+	fn on_listen_for_chunks_received(
+		&mut self,
+		runtime_handle: &mut Handle,
+		sender: &mut mpsc::UnboundedSender<WorkerMsg>,
+		relay_parent: Hash,
+		candidate_hash: Hash,
+		id: usize
+	) -> Result<(), Error> {
+		let candidate = self.availability_store.get_candidate(&candidate_hash)
+			.ok_or(Error::CandidateNotFound { candidate_hash })?;
+
+		if self.availability_store
+			.get_erasure_chunk(&relay_parent, candidate.block_data_hash, id)
+			.is_none() {
+			if let Err(e) = self.register_chunks_listener(
+				runtime_handle,
+				sender,
+				relay_parent,
+				candidate.erasure_root
+			) {
+				warn!(target: LOG_TARGET, "Failed to register a gossip listener: {}", e);
+			}
+		}
+
+		Ok(())
+	}
+
+	pub fn start(mut self) -> WorkerHandle {
+		let sender = self.sender.clone();
+		let mut receiver = self.receiver.take()
+			.expect("Should start a worker with a receiver side; qed");
+
+		let (signal, exit) = exit_future::signal();
+
+		let handle = thread::spawn(move || {
+			let mut sender = self.sender.clone();
+			let mut runtime = LocalRuntime::new().expect("Could not create local runtime");
```

Comment:
This seems like it could fail. Panickers should be proven not to fail or removed.

Reply:
This just copies [logic](https://github.com/paritytech/polkadot/blob/master/validation/src/attestation_service.rs#L141) in `attestation_service`, should we just fail with an error here before starting a thread and return that to the caller?

Reply:
If possible, that seems reasonable. Is the `LocalRuntime` `Send`? We could create and then send it if so

Reply:
No, only a handle to it is